src/main/java/hudson/plugins/git/GitSCM.java

```diff
@@ -1345,13 +1349,44 @@ private void computeChangeLog(GitClient git, Revision revToBuild, TaskListener l
             }
 
             if (!exclusion) {
-                // this is the first time we are building this branch, so there's no base line to compare against.
-                // if we force the changelog, it'll contain all the changes in the repo, which is not what we want.
-                listener.getLogger().println("First time build. Skipping changelog.");
-            } else {
-                changelog.to(out).max(MAX_CHANGELOG).execute();
-                executed = true;
+                listener.getLogger().println("First build of this branch, computing origin.");
+                // Ok, so here we are naked ... All we have is a detached head, on a SHA1 checkout, with no informations on remote heads/refs
+                // So we first list all known SHA1's from jobs having at least one remote url in common
+                // Then we go down current rev's ancestors until we found one already built, this is our stop
+
+                Set<String> ourUrls = userRemoteConfigs.stream().map(c -> c.getUrl()).collect(Collectors.toSet());
+
+                Instant start = Instant.now();
+                List<ObjectId> knownSHA1 = (List<ObjectId>) Jenkins.get().getItems().stream()
```

Comment:
I don't think that it is safe for performance reasons to iterate over all the builds in the system to find a possible predecessor SHA1.  Some Jenkins servers may have many millions of builds in the system and will suffer serious performance issues while this code is iterating through all the builds.

@rsandell or @fcojfernandez may have more insights to offer on ways that a candidate predecessor could be identified without requiring that the plugin iterate over every build in the system.

Reply:
I totally agree and I would love to find a better approach.
Just for information, on our production server, 1300 jobs, 7000 builds and 100 000 actions, it take 200ms to loop over all.

Reply:
Yea this is really bad. If it only takes 200ms to loop over all builds in your system that probably means you have all in memory and perhaps few complex pipeline builds that has to load from disk whenever you try to access one.

@jglick has, floating around in Jira somewhere, an idea for an API to find a previous candidate build for for example multibranch builds where the previous revision for a PR build might be in the last master branch build for example.

Also since we are hoping to get rid of the bloated BuildData as soon as we are able perhaps it's not wise to rely too much on it? Though at this point it's the only source of data we have I think.

Reply:
One potential way to lessen the load would be to only search through the current job's parent or grand parent. But I think even that would be a hard no from me.

Reply:
A possible solution that might be feasible would be to store the commits built in a central (to the master) database, like SQLite or H2, when a changelog is generated and then later look it up from there, pick the build who's job has the closest similar path to the job.
Some database maintenance will be required though, like an ItemListener renaming jobs when a job is relocated and removing build references when builds and jobs are being removed etc.

Reply:
Ok, totally understand and agree with your case. I will try the database way. Would you mind explain which informations you want in it (remote url, branch names, sha1, ... ?)

Reply:
> @jglick has, floating around in Jira somewhere, an idea for an API…

[JENKINS-33274](https://issues.jenkins-ci.org/browse/JENKINS-33274).

Not sure I understand what this PR is trying to do, but it looks like an unworkable approach. I do not think any database is needed, either. This issue is not something to solve in isolation in `git-plugin`; it requires knowledge of project structure, meaning some kind of API in `scm-api` and/or `branch-api` plus implementations in SCM plugins plus usage in `workflow-multibranch`.

Reply:
Ok.

This PR is trying to solve JENKINS-14138, which is exactly the same as JENKINS-33274, but dedicated to git-plugin : when jenkins build a new, freshly pushed, branch from a git repository, changelog is empty.

I just discover JENKINS-33274, and I agree that this PR could just be the "major SCM plugins needs to implement this API" part of JENKINS-33274, for Git - at least for multibranch pipelines. For other kinds of projects, it has to be solved another way

Reply:
> explain which informations you want in it (remote url, branch names, sha1, ... ?)

Whatever is needed to get the job done I guess :)
Normalised repo url, sha, job-path, build's externalizable id. Are the minimum I think.
Question is if we should store all commits that are new in the build or just the current HEAD?
HEAD might be enough, or in risk of missing some builds maybe all shas in the changelog with a mark on which one is the head? 

Or maybe forget about the database for now and try to tackle JENKINS-33274 :)

Reply:
> Or maybe forget about the database for now and try to tackle JENKINS-33274 :)

whatever you want, but JENKINS-33274 will limit the fix to multibranch pipelines, or maybe all pipelines no ? So other kind of projects (freestyle, maven, ...) will still not work. But maybe, pipelines are the standard now ?

Reply:
If you are using SCM variable substitution tricks to switch the branch from build to build inside a single `Job` (traditional or Pipeline), it is not even clear to me what the behavior _should_ be. The point of multibranch in this context is not that it is Pipeline (in fact there are some basically abandoned plugins that implement multibranch for other job types), but that it defines a stronger model for where commits of particular branches are built—which is structure that we could use here.